File: 210-course-schedule-ii/210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph=defaultdict(list)
        degree=[0]*(numCourses)
        for i,j in prerequisites:
            graph[j].append(i)
            degree[i]+=1
        queue=deque()    
        ordering=[]
        for i in range(numCourses):
            if degree[i]==0:
                queue.append(i)
                ordering.append(i)
        while queue:
            front=queue.popleft()
            for out in graph[front]:
                degree[out]-=1
                if degree[out]==0:
                    queue.append(out)
                    ordering.append(out)
        if sum(degree)==0:
            return ordering
        else:
            return []
        
        

File: 210-course-schedule-ii/210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        
        vector<int> degree(numCourses);
        vector<int> adj[numCourses];
        for(auto x:prerequisites){
            adj[x[1]].push_back(x[0]);
            degree[x[0]]++;
        }
        queue<int> q;
        vector<int> ans;
        for(int i=0;i<numCourses;i++){
            if(degree[i]==0){
                q.push(i);
                ans.push_back(i);
            }
        }
        while(!q.empty()){
            int aux=q.front();
            q.pop();
            for(auto out:adj[aux]){
                degree[out]--;
                if(degree[out]==0){
                    q.push(out);
                    ans.push_back(out);
                }
            }
        }
        //return ans;
        if(ans.size()==numCourses){
            return ans;
        }
        return {};      
    }
};